### PR TITLE
[core] move parallel sampling out from vllm core

### PR DESCRIPTION
**Pull request by youkaichao**
try to hide seq group from the core, by handling parallel sampling in llm engine.

### CODE REVIEW

**Comment by youkaichao** [comment]
the caveat is that we cannot support streaming when `n > 1` . I think people don't use streaming when `n > 1`, and it is not clearly defined.

say we have n = 5, and the first stream gives 5 tokens, and then sequence 2 finish, do we send 5 outputs with the 2nd as empty? or send 4 outputs and let users mantain the status?

**Comment by youkaichao** [comment]
@robertgshaw2-neuralmagic  can you help take a look? I met a strange error:

`pytest -v -s tests/entrypoints/openai/test_completion.py::test_guided_json_completion[-outlines]`

will fail in this implementation. `lm-format-enforcer` works well, and `--disable-frontend-multiprocessing` also works. only the combination of the mqllmengine and outlines does not work.

File: vllm/engine/llm_engine.py [hunk 1]
```diff
@@ -474,6 +475,8 @@ def get_tokenizer_for_seq(sequence: Sequence) -> AnyTokenizer:
                 ),
             ))
 
+        self.group_id_to_holders: Dict[str, SeqGroupHolder] = {}
```
Comment (zhuohan123) [inline vllm/engine/llm_engine.py:478]: Probably a better name?
```suggestion
        self.request_id_to_group_holders: Dict[str, SeqGroupHolder] = {}
```

File: vllm/outputs.py [hunk 2]
```diff
@@ -126,6 +128,18 @@ def from_seq_group(cls, seq_group: SequenceGroup,
                 not finished):
             return None
 
+        if finished and seq_group.request_id in group_id_to_holders:
+            # parallel sampling can not stream the output
```
Comment (zhuohan123) [inline vllm/outputs.py:132]: Why is this the case? Theoretically for parallel sampling I think we can just concurrently stream multiple outputs back. We can interleave the output streams of different samples.

**Review by zhuohan123 — APPROVED** [review]

**Comment by afeldman-nm** [comment]
@youkaichao how would this PR impact `best_of > 1` requests? Is `best_of` functionality still within the engine, or is it moved outside the engine as has been done for beam search?

**Comment by youkaichao** [comment]
@afeldman-nm `best_of > 1` is already converted to parallel sampling in https://github.com/vllm-project/vllm/pull/9261/

File: vllm/engine/llm_engine.py (hunk 1 above)
Comment (youkaichao) [reply vllm/engine/llm_engine.py]: fixed in [245920f](https://github.com/vllm-project/vllm/pull/9302/commits/245920f17ae10675d98a4e5fc92a0e326cb3fc93)

File: vllm/outputs.py (hunk 2 above)
Comment (youkaichao) [reply vllm/outputs.py]: fixed in [245920f](https://github.com/vllm-project/vllm/pull/9302/commits/245920f17ae10675d98a4e5fc92a0e326cb3fc93)

**Comment by rand-fly** [comment]
Could you explain the benefit of doing so? It seems that with this change, the scheduler can no longer make decisions based on the number of sequences within a `SequenceGroup`.

**Comment by youkaichao** [comment]
> Could you explain the benefit of doing so? It seems that with this change, the scheduler can no longer make decisions based on the number of sequences within a `SequenceGroup`.

yes, the scheduler will only process single sequence in the future, to make the core code simple.

**Comment by rand-fly** [comment]
This modification makes the "fork" mechanism of vLLM completely unused. Previously, for a request with n > 1, its prompt was prefilled only once, and then the sequence was "forked" into n sequences to avoid redundant computation. After this modification, a request with n > 1 has to prefill its prompt n times.
A small experiment code can be used to verify this. Notice how it has become much slower after this squashed commit.

```py
from vllm import LLM, SamplingParams
import time

# Sample prompts.
prompts = [
    "Once upon a time, there was a king.",
]
# Create a sampling params object.
sampling_params = SamplingParams(seed=42, temperature=0.1, max_tokens=1, n=100)

# Create an LLM.
llm = LLM(model="Qwen/Qwen2.5-1.5B-Instruct")

# warm up
outputs = llm.generate(prompts, sampling_params) 

begin_time = time.time()
outputs = llm.generate(prompts, sampling_params)
end_time = time.time()

print(f"{end_time - begin_time}s")

# Print the outputs.
for output in outputs:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")
```

**Comment by youkaichao** [comment]
> This modification makes the "fork" mechanism of vLLM completely unused. Previously, for a request with n > 1, its prompt was prefilled only once, and then the sequence was "forked" into n sequences to avoid redundant computation. After this modification, a request with n > 1 has to prefill its prompt n times.

Yes, this is intended. Please use prefix caching to speed up and share the prefill. All the sharing will not be hardcoded in the scheduler, and will only happen through prefix caching.

I'm not sure if prefix caching currently supports sharing in the same batch. If you want optimal performance, I would suggest running a `n=1` request first, and then run another `n=n-1` request.

**Comment by rand-fly** [comment]
> > This modification makes the "fork" mechanism of vLLM completely unused. Previously, for a request with n > 1, its prompt was prefilled only once, and then the sequence was "forked" into n sequences to avoid redundant computation. After this modification, a request with n > 1 has to prefill its prompt n times.
> 
> Yes, this is intended. Please use prefix caching to speed up and share the prefill. All the sharing will not be hardcoded in the scheduler, and will only happen through prefix caching.
> 
> I'm not sure if prefix caching currently supports sharing in the same batch. If you want optimal performance, I would suggest running a `n=1` request first, and then run another `n=n-1` request.

Thank you for clarifying. Prefix caching does support sharing in the same batch, though the performance gain is not as much as using the "fork" mechanism.

**Pull request merged**